src/modules/mavlink/mavlink_messages.cpp

```diff
@@ -253,242 +253,105 @@ static_assert(MAV_SENSOR_ROTATION_ROLL_90_PITCH_68_YAW_293 == static_cast<MAV_SE
+		custom_mode.main_mode = PX4_CUSTOM_MAIN_MODE_ACRO;
+		break;
 
-	/* fallthrough */
 	case vehicle_status_s::NAVIGATION_STATE_DESCEND:
-		*mavlink_base_mode |= auto_mode_flags;
-		custom_mode->main_mode = PX4_CUSTOM_MAIN_MODE_AUTO;
-		custom_mode->sub_mode = PX4_CUSTOM_SUB_MODE_AUTO_LAND;
+		custom_mode.main_mode = PX4_CUSTOM_MAIN_MODE_AUTO;
+		custom_mode.sub_mode = PX4_CUSTOM_SUB_MODE_AUTO_LAND;
 		break;
 
 	case vehicle_status_s::NAVIGATION_STATE_TERMINATION:
-		*mavlink_base_mode |= MAV_MODE_FLAG_MANUAL_INPUT_ENABLED;
-		custom_mode->main_mode = PX4_CUSTOM_MAIN_MODE_MANUAL;
+		custom_mode.main_mode = PX4_CUSTOM_MAIN_MODE_MANUAL;
 		break;
 
 	case vehicle_status_s::NAVIGATION_STATE_OFFBOARD:
-		*mavlink_base_mode |= auto_mode_flags;
-		custom_mode->main_mode = PX4_CUSTOM_MAIN_MODE_OFFBOARD;
+		custom_mode.main_mode = PX4_CUSTOM_MAIN_MODE_OFFBOARD;
 		break;
 
-	case vehicle_status_s::NAVIGATION_STATE_MAX:
-		/* this is an unused case, ignore */
+	case vehicle_status_s::NAVIGATION_STATE_STAB:
+		custom_mode.main_mode = PX4_CUSTOM_MAIN_MODE_STABILIZED;
 		break;
 
-	}
-}
-
-static void get_mavlink_mode_state(const struct vehicle_status_s *const status, uint8_t *mavlink_state,
-				   uint8_t *mavlink_base_mode, uint32_t *mavlink_custom_mode)
-{
-	*mavlink_state = 0;
-	*mavlink_base_mode = 0;
-	*mavlink_custom_mode = 0;
-
-	union px4_custom_mode custom_mode;
-	get_mavlink_navigation_mode(status, mavlink_base_mode, &custom_mode);
-	*mavlink_custom_mode = custom_mode.data;
-
-	/* set system state */
-	if (status->arming_state == vehicle_status_s::ARMING_STATE_INIT
-	    || status->arming_state == vehicle_status_s::ARMING_STATE_IN_AIR_RESTORE
-	    || status->arming_state == vehicle_status_s::ARMING_STATE_STANDBY_ERROR) {	// TODO review
-		*mavlink_state = MAV_STATE_UNINIT;
-
-	} else if (status->arming_state == vehicle_status_s::ARMING_STATE_ARMED) {
-		*mavlink_state = MAV_STATE_ACTIVE;
-
-	} else if (status->arming_state == vehicle_status_s::ARMING_STATE_STANDBY) {
-		*mavlink_state = MAV_STATE_STANDBY;
-
-	} else if (status->arming_state == vehicle_status_s::ARMING_STATE_SHUTDOWN) {
-		*mavlink_state = MAV_STATE_POWEROFF;
-
-	} else {
-		*mavlink_state = MAV_STATE_CRITICAL;
-	}
-}
-
-
-class MavlinkStreamHeartbeat : public MavlinkStream
-{
-public:
-	const char *get_name() const override
-	{
-		return MavlinkStreamHeartbeat::get_name_static();
-	}
-
-	static constexpr const char *get_name_static()
-	{
-		return "HEARTBEAT";
-	}
+	case vehicle_status_s::NAVIGATION_STATE_RATTITUDE:
+		custom_mode.main_mode = PX4_CUSTOM_MAIN_MODE_RATTITUDE;
+		break;
 
-	static constexpr uint16_t get_id_static()
-	{
-		return MAVLINK_MSG_ID_HEARTBEAT;
-	}
+	case vehicle_status_s::NAVIGATION_STATE_AUTO_TAKEOFF:
+		custom_mode.main_mode = PX4_CUSTOM_MAIN_MODE_AUTO;
+		custom_mode.sub_mode = PX4_CUSTOM_SUB_MODE_AUTO_TAKEOFF;
+		break;
 
-	uint16_t get_id() override
-	{
-		return get_id_static();
-	}
+	case vehicle_status_s::NAVIGATION_STATE_AUTO_LAND:
```

Comment:
I thought fallthrough for the land modes was clear and more readable but either works.

Reply:
I was trying to keep it 1-1 with the `vehicle_status_s::NAVIGATION_STATE` ordering to not miss anything. I suspect this kind of mapping is going to get more complicated in the future with many more internal PX4 flight modes (evolved flight_tasks) that aren't individually exposed.